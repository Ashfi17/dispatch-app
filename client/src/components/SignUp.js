import React, { Component } from "react";

import "../assets/styles/login.css";

import { createUser } from "../actions/Users";

import { connect } from "react-redux";

export class SignUp extends Component {
  state = {
    name: "",
    email: "",
    address: "",
    password: "",
    confirmPassword: "",
    showError: false,
    errorMessage: "",
  };

  OnTextChange = (event) => {
    this.setState({
      errorMessage: "",
      [event.target.name]: event.target.value,
    });
  };

  validateAndRegister = async () => {
    if (
      !this.state.name ||
      !this.state.email ||
      !this.state.password ||
      !this.state.confirmPassword
    ) {
      this.setState({
        errorMessage: "Enter all the fields",
      });
    } else {
      if (this.state.password !== this.state.confirmPassword) {
        this.setState({
          errorMessage: "Password does not match",
        });
      } else {
        if (this.state.password.length < 5) {
          this.setState({
            errorMessage: "Password length should be greater then 5!",
          });
        } else {
          if (this.state.email.length < 5) {
            this.setState({
              errorMessage: "Please enter valid email!",
            });
          } else {
            this.setState({
              errorMessage: "",
            });
            let user = {
              user_name: this.state.name,
              user_email: this.state.email,
              user_password: this.state.password,
              user_address: this.state.address,
            };
            await this.props.createUser(user, this.props.history);

            this.props.errorMessage
              ? this.setState({ errorMessage: this.props.errorMessage })
              : this.setState({
                  name: "",
                  email: "",
                  password: "",
                  address: "",
                  confirmPassword: "",
                });
          }
        }
      }
    }
  };

  render() {
    return (
      <div>
        <div className="login-signup-card">
          <h3 className="login-Signup-Text-header">SignUp</h3>
          <div className="login-signup-elements-container">
            <input
              id="signup-name"
              className="input-field"
              name="name"
              value={this.state.name}
              placeholder="enter name"
              onChange={(e) => this.OnTextChange(e)}
            />
            <input
              id="signup-email"
              className="input-field"
              name="email"
              placeholder="enter email"
              value={this.state.email}
              onChange={(e) => this.OnTextChange(e)}
            />
            <input
              id="signup-address"
              className="input-field"
              name="address"
              placeholder="enter address"
              value={this.state.address}
              onChange={(e) => this.OnTextChange(e)}
            />
            <input
              id="signup-password"
              className="input-field"
              name="password"
              placeholder="enter password"
              type="password"
              value={this.state.password}
              onChange={(e) => this.OnTextChange(e)}
            />
            <input
              id="signup-confirm-password"
              className="input-field"
              name="confirmPassword"
              placeholder="confirm password"
              type="password"
              value={this.state.confirmPassword}
              onChange={(e) => this.OnTextChange(e)}
            />
            <span className="error-text">{this.state.errorMessage}</span>
            <button
              id="signup-submit-button"
              className="submit-button"
              onClick={this.validateAndRegister}
            >
              SignUp
            </button>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  createUserMessage: state.UserReducer.createUserMessage,
  errorMessage: state.UserReducer.errorMessage,
});

export default connect(mapStateToProps, { createUser })(SignUp);
