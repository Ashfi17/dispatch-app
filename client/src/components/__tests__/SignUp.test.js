import React from "react";
import { shallow } from "enzyme";
import { SignUp } from "../SignUp";

const createUser = jest.fn();
const history = { push: jest.fn() };

const wrapper = shallow(<SignUp createUser={createUser} history={history} />);

describe("Signup test suite", () => {
  it("should match the snapshpot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should check if the text is being typed and the onTextChange function is called", () => {
    const OnTextChange = jest.spyOn(wrapper.instance(), "OnTextChange");
    const name = {
      target: "dummy name",
    };
    const email = {
      target: "dummy email",
    };

    const password = {
      target: "dummy password",
    };
    wrapper.find("#signup-name").simulate("change", name);
    expect(OnTextChange).toBeCalled();
    wrapper.find("#signup-email").simulate("change", email);
    expect(OnTextChange).toBeCalled();

    wrapper.find("#signup-password").simulate("change", password);
    expect(OnTextChange).toBeCalled();
  });
  it("shoud check if the condtions for register are being validated and if not it should show the error message", () => {
    wrapper.setState({
      name: "test name",
      email: "",
      address: "address123",
      password: "",
      confirmPassword: "",
    });
    wrapper.find("#signup-submit-button").simulate("click");
    expect(wrapper.state().errorMessage).toBe("Enter all the fields");
    wrapper.setState({
      name: "test name",
      email: "dummy@gmail.com",
      address: "address line 123",
      password: "abcdefg",
      confirmPassword: "abcfgt565",
    });
    wrapper.find("#signup-submit-button").simulate("click");
    expect(wrapper.state().errorMessage).toBe("Password does not match");
    wrapper.setState({
      name: "test name",
      email: "dummy@gmail.com",
      address: "address line 123",
      password: "abc",
      confirmPassword: "abc",
    });
    wrapper.find("#signup-submit-button").simulate("click");
    expect(wrapper.state().errorMessage).toBe(
      "Password length should be greater then 5!"
    );
  });
  it("should check if the action is called if all the condtions are validated", () => {
    const validateAndRegister = jest.spyOn(
      wrapper.instance(),
      "validateAndRegister"
    );
    wrapper.setState({
      name: "test name",
      email: "dummy@gmail.com",

      password: "abc1234",
      confirmPassword: "abc1234",
    });
    wrapper.find("#signup-submit-button").simulate("click");
    expect(validateAndRegister).toBeCalled();
    expect(wrapper.state().errorMessage).toBe("");
    expect(createUser).toBeCalledWith(
      {
        user_name: "test name",
        user_email: "dummy@gmail.com",

        user_password: "abc1234",
      },
      history
    );
  });
});
