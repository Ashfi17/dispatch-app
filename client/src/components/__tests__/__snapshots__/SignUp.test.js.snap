// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup test suite should match the snapshpot 1`] = `
<div>
  <div
    className="login-signup-card"
  >
    <h3
      className="login-Signup-Text-header"
    >
      SignUp
    </h3>
    <div
      className="login-signup-elements-container"
    >
      <input
        className="input-field"
        id="signup-name"
        name="name"
        onChange={[Function]}
        placeholder="enter name"
        value=""
      />
      <input
        className="input-field"
        id="signup-email"
        name="email"
        onChange={[Function]}
        placeholder="enter email"
        value=""
      />
      <input
        className="input-field"
        id="signup-address"
        name="address"
        onChange={[Function]}
        placeholder="enter address"
        value=""
      />
      <input
        className="input-field"
        id="signup-password"
        name="password"
        onChange={[Function]}
        placeholder="enter password"
        type="password"
        value=""
      />
      <input
        className="input-field"
        id="signup-confirm-password"
        name="confirmPassword"
        onChange={[Function]}
        placeholder="confirm password"
        type="password"
        value=""
      />
      <span
        className="error-text"
      />
      <button
        className="submit-button"
        id="signup-submit-button"
        onClick={[Function]}
      >
        SignUp
      </button>
    </div>
  </div>
</div>
`;
